### DISCUSSION TITLE
Recursive XML-like tags

### DISCUSSION THREAD

**Discussion by soywod** [zesterer/chumsky#224] [Q&A] [answered]
Here my main struct:

```rust

#[derive(Clone, Debug, Eq, PartialEq)]
pub(crate) enum Part {
    MultiPart(Props, Vec<Part>),
    TextPlainPart(String),
}
```

A multi part is just a XML-like tag that can contains other multi parts `<#multipart>…<#/multipart>`. A text plain part is everything else. Here my parser:

```rust
fn multipart() -> impl Parser<char, Part, Error = Simple<char>> {
    recursive(|multipart| {
        just("<#multipart>")
            .padded()
            .ignore_then(
                take_until(choice((
                    just("<#multipart>")
                        .padded()
                        .rewind()
                        .ignore_then(multipart.map(|m| vec![m])),
                    just("<#/multipart>").padded().map(|_| vec![]),
                )))
                .map(|(chars, mut parts)| {
                    if !chars.is_empty() {
                        parts.insert(0, Part::TextPlainPart(chars.iter().collect()))
                    }
                    parts
                })
                .repeated()
                .flatten(),
            )
            .map(|children| Part::MultiPart(HashMap::default(), children))
    })
}
```

The recursion works well (tests pass), but I have trouble when multi parts are next to each other (siblings / adjacent). Here my test:

```rust
#[test]
fn test_adjacent_multiparts() {
    assert_eq!(
        multipart().parse(concat_line!(
            "<#multipart>",
            "<#multipart>",
            "a",
            "<#/multipart>",
            "<#multipart>",
            "b",
            "<#/multipart>",
            "<#/multipart>",
        )),
        Ok(Part::MultiPart(
            HashMap::default(),
            vec![
                Part::MultiPart(
                    HashMap::default(),
                    vec![Part::TextPlainPart(String::from("a"))]
                ),
                Part::MultiPart(
                    HashMap::default(),
                    vec![Part::TextPlainPart(String::from("b"))]
                )
            ]
        )),
    );
}

// running 1 test
// test parsers::test::test_adjacent_multiparts ... FAILED
// 
// failures:
// 
// ---- parsers::test::test_adjacent_multiparts stdout ----
// thread 'parsers::test::test_adjacent_multiparts' panicked at 'assertion failed: `(left == right)`
//   left: `Ok(MultiPart({}, [MultiPart({}, [TextPlainPart("a"), MultiPart({}, [TextPlainPart("b")])])]))`,
//  right: `Ok(MultiPart({}, [MultiPart({}, [TextPlainPart("a")]), MultiPart({}, [TextPlainPart("b")])]))`', src/parsers.rs:430:9
// note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

It looks like the last multi part is included inside the previous one, instead of being at the same level. I cannot find what is wrong there, any idea?

**Comment by soywod** [comment, marked as answer]
Ok I was able to make the test pass with another implementation:

```rust

fn multipart() -> impl Parser<char, Part, Error = Simple<char>> {
    recursive(|multipart| {
        choice((
            multipart,
            not_multipart_end()
                .repeated()
                .at_least(1)
                .collect()
                .map(Part::TextPlainPart),
        ))
        .repeated()
        .delimited_by(just("<#multipart>"), just("<#/multipart>"))
        .map(|children| Part::MultiPart(HashMap::default(), children))
    })
}
```

The problem is the implementation of `not_multipart_end`, which is quite… ugly. If you have a better option let me know:

```rust

fn not_multipart_end() -> impl Parser<char, char, Error = Simple<char>> {
    none_of('<')
        .then_ignore(none_of('#').or_not().rewind())
        .then_ignore(none_of('/').or_not().rewind())
        .then_ignore(none_of('m').or_not().rewind())
        .then_ignore(none_of('u').or_not().rewind())
        .then_ignore(none_of('l').or_not().rewind())
        .then_ignore(none_of('t').or_not().rewind())
        .then_ignore(none_of('i').or_not().rewind())
        .then_ignore(none_of('p').or_not().rewind())
        .then_ignore(none_of('a').or_not().rewind())
        .then_ignore(none_of('r').or_not().rewind())
        .then_ignore(none_of('t').or_not().rewind())
        .then_ignore(none_of('>').or_not().rewind())
}
```

**Reply by soywod** [reply]
In fact it does not work, `not_multipart_end` fails to parse string like `a is > than b` because of the `>` :slightly_frowning_face:

[edit] sth like this seems to work but I cannot believe it is the only option available over there:

```rust
fn any_string_except_multipart_end_tag() -> impl Parser<char, char, Error = Simple<char>> {
    none_of('<').or(
        just('<').then_ignore(none_of('#').or(just('#').then_ignore(none_of('/').or(
            just('/').then_ignore(none_of('m').or(just('m').then_ignore(none_of('u').or(
                just('u').then_ignore(none_of('l').or(just('l').then_ignore(none_of('t').or(
                    just('t').then_ignore(none_of('i').or(just('i').then_ignore(none_of('p').or(
                        just('p').then_ignore(none_of('a').or(just('a').then_ignore(
                            none_of('r').or(just('r').then_ignore(
                                none_of('t').or(just('t').then_ignore(none_of('>').or(just('>')))),
                            )),
                        ))),
                    )))),
                )))),
            )))),
        )))),
    )
}
```

**Reply by zesterer** [reply]
Would the new `not` combinator on `master` help here? `a.not()` parses any token that *doesn't* match `a`.

**Reply by soywod** [reply]
It could help, but I cannot make the compiler happy:

```rust
fn any_string_except_part_tags() -> impl Parser<char, Part, Error = Simple<char>> {
    choice((
        just("<#multipart"),
        just("<#/multipart>"),
        just("<#part"),
        just("<#/part>"),
    ))
    .not()
    .map(ToOwned::to_owned)
    .map(Part::TextPlainPart)
}
```

I get sth like:

```
error[E0277]: the trait bound `chumsky::combinator::Not<Choice<(Just<&str, &str, _>, Just<&str, &str, _>, Just<&str, &str, _>, Just<&str, &str, _>), _>, &str>: chumsky::Parser<char, &str>` is not satisfied
  --> src/parsers.rs:26:41
   |
26 | fn any_string_except_part_tags<'a>() -> impl Parser<char, Part, Error = Simple<char>> {
   |                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `chumsky::Parser<char, &str>` is not implemented for `chumsky::combinator::Not<Choice<(Just<&str, &str, _>, Just<&str, &str, _>, Just<&str, &str, _>, Just<&str, &str, _>), _>, &str>`
   |
   = help: the trait `chumsky::Parser<I, I>` is implemented for `chumsky::combinator::Not<A, O>`
   = note: required because of the requirements on the impl of `chumsky::Parser<char, std::string::String>` for `chumsky::combinator::Map<chumsky::combinator::Not<Choice<(Just<&str, &str, _>, Just<&str, &str, _>, Just<&str, &str, _>, Just<&str, &str, _>), _>, &str>, for<'r> fn(&'r str) -> <str as ToOwned>::Owned {<str as ToOwned>::to_owned}, &str>`
   = note: 1 redundant requirement hidden
   = note: required because of the requirements on the impl of `chumsky::Parser<char, part::Part>` for `chumsky::combinator::Map<chumsky::combinator::Map<chumsky::combinator::Not<Choice<(Just<&str, &str, _>, Just<&str, &str, _>, Just<&str, &str, _>, Just<&str, &str, _>), _>, &str>, for<'r> fn(&'r str) -> <str as ToOwned>::Owned {<str as ToOwned>::to_owned}, &str>, fn(std::string::String) -> part::Part {part::Part::TextPlainPart}, std::string::String>`
```

I was able to get rid of this error by changing the signature of the function:

```rust
fn any_string_except_part_tags<'a>() -> impl Parser<&'static str, Part, Error = Simple<&'static str>>
```

But then I cannot use it because it has a different `I` type than the rest of the parsers (`char` vs `&str`). How to make them coexist?

**Reply by zesterer** [reply]
`.not()` will only parse a single character at maximum. If you want a whole, well, string of characters that aren't the start of the given pattern, `choice(...).not().repeated().collect::<String>()` is the way to go.

**Reply by soywod** [reply]
Seems to works well, the final version looks like this:

```rust

fn any_string_except_part_tags() -> impl Parser<char, Part, Error = Simple<char>> {
    choice((
        just("<#multipart"),
        just("<#/multipart>"),
        just("<#part"),
        just("<#/part>"),
    ))
    .padded()
    .rewind()
    .not()
    .repeated()
    .at_least(1)
    .collect::<String>()
    .map(Part::TextPlainPart)
}
```

All my [tests](https://git.sr.ht/~soywod/mime-tpl-lib/tree/master/item/src/parsers.rs#L427-517) pass, I can go forward now. Thanks a lot for your help :pray:

**Reply by zesterer** [reply]
Nice! Fwiw, that `.rewind()` isn't doing anything, I think. Does it still work fine if removed?

**Reply by soywod** [reply]
Indeed it works without. So it means the `.not` operator does not consume the input, right? Good to know!

**Reply by zesterer** [reply]
`.not()` *does* consume input, just only ever a single character on success.

Note that this behaviour will eventually be changing when #9 is merged and you'll need to do `any().and_is(a.not())`, but that won't be for a while and there will be a new release before then anyway.

**Reply by soywod** [reply]
Thank you for the information. Just a question: when do you plan to release a version that will contain this `.not()` operator? No pressure of course, it is just to known when I will be able to release my lib, I'm stuck for now:

```sh
$ cargo publish
error: all dependencies must have a version specified when publishing.
dependency `chumsky` does not specify a version
Note: The published dependency will use the version from crates.io,
the `git` specification will be removed from the dependency declaration.
```

**Reply by zesterer** [reply]
When `zero-copy` is finished, I plan to do one last release of the current master branch before it's merged. For now I'm holding fire on that, but that position is starting to become somewhat untenable given the number of changes that have occurred. Is this a blocking issue for you right now?

**Reply by soywod** [reply]
I'm actually working with your parser on [mime-msg-builder](https://git.sr.ht/~soywod/mime-msg-builder), which will be added to my main lib [himalaya-lib](https://git.sr.ht/~soywod/himalaya-lib) (here I need a version to publish), then I will be able to update the main project [himalaya](https://github.com/soywod/himalaya). This last one is a binary so I do not need to publish it on `crates.io`, so in theory I can stick to git+rev for a while. So no pressure :slightly_smiling_face: I will let you know if someday it blocks me a lot.

**Reply by soywod** [reply]
@zesterer in fact Nix builds are blocked on the CLI because of missing a version, I thought it would work without be it seems not :slightly_frowning_face: so I can tell now that it is a blocking issue. Let me know whenever a release is ready from your side so I can release mine!

**Reply by zesterer** [reply]
I'm going to try to find the time to run a release either tomorrow or Tuesday.

**Reply by zesterer** [reply]
(I have not forgotten about this, just yet to find an adequate period of free time)

**Reply by soywod** [reply]
Don't worry all is fine, I will wait as much time as you need, no pressure!

**Reply by soywod** [reply]
Is there anything I could do to help releasing a next version?

**Reply by zesterer** [reply]
Hello @soywod! It's been a long time in coming, but I just released 0.9: https://github.com/zesterer/chumsky/releases/tag/0.9

**Reply by apps4uco** [reply]
Hi, I made a simplified version of your code

> 
> ```rust
> fn multipart() -> impl Parser<char, Part, Error = Simple<char>> {
>     recursive(|multipart| {
>         choice((
>             multipart,
>             not_multipart_end()
>                 .repeated()
>                 .at_least(1)
>                 .collect()
>                 .map(Part::TextPlainPart),
>         ))
>         .repeated()
>         .delimited_by(just("<#multipart>"), just("<#/multipart>"))
>         .map(|children| Part::MultiPart(HashMap::default(), children))
>     })
> }
> ```
> 
> The problem is the implementation of `not_multipart_end`, which is quite… ugly. If you have a better option let me know:
> 
> ```rust
> fn not_multipart_end() -> impl Parser<char, char, Error = Simple<char>> {
>     none_of('<')
>         .then_ignore(none_of('#').or_not().rewind())
>         .then_ignore(none_of('/').or_not().rewind())
>         .then_ignore(none_of('m').or_not().rewind())
>         .then_ignore(none_of('u').or_not().rewind())
>         .then_ignore(none_of('l').or_not().rewind())
>         .then_ignore(none_of('t').or_not().rewind())
>         .then_ignore(none_of('i').or_not().rewind())
>         .then_ignore(none_of('p').or_not().rewind())
>         .then_ignore(none_of('a').or_not().rewind())
>         .then_ignore(none_of('r').or_not().rewind())
>         .then_ignore(none_of('t').or_not().rewind())
>         .then_ignore(none_of('>').or_not().rewind())
> }
> ```


I replaced the not_multipart_end() function with

any().and_is(just("<#/multipart>").not())

Im not an expert in chumsky so I dont know if there is any performance impact, but it passes the test case ok and is much easier to understand/write. Hope it helps.
         
```
use chumsky::prelude::*;

#[derive(Clone, Debug, Eq, PartialEq)]
pub(crate) enum Part {
    MultiPart(Vec<Part>),
    TextPlainPart(String),
}

fn multipart<'a>() -> impl Parser<'a, &'a str, Part, extra::Err<Rich<'a, char>>> {
    recursive(|multipart| {
        choice((
            multipart,
            any()
                .and_is(just("<#/multipart>").not())
                .repeated()
                .at_least(1)
                .collect::<String>()
                .map(Part::TextPlainPart),
        ))
        .repeated()
        .collect::<Vec<_>>()
        .delimited_by(just("<#multipart>"), just("<#/multipart>"))
        .map(|children| Part::MultiPart(children))
    })
}

#[test]
fn test_adjacent_multiparts() {
    let input = "<#multipart><#multipart>a<#/multipart><#multipart>b<#/multipart><#/multipart>";
    assert_eq!(
        multipart().parse(input).into_result(),
        Ok(Part::MultiPart(vec![
            Part::MultiPart(vec![Part::TextPlainPart(String::from("a"))]),
            Part::MultiPart(vec![Part::TextPlainPart(String::from("b"))])
        ])),
    );
}
```
